### ISSUE TITLE
Bug: Plugins are "ignored" when provider name is variable

### ISSUE THREAD

**Issue by MacMcIrish** [serverless/serverless#4837]
# This is a Bug Report

## Description
When the provider name is a variable, the proper plugin may not be used.

For bug reports:
* What went wrong?
The loaded `deploy` plugin was lib/plugins/deploy.
* What did you expect should have happened?
The loaded `deploy` plugin to be lib/plugins/aws/common/deploy
* What was the config you used?
```
---
custom:
  name: aws

provider:
  name: ${self:custom.name}
---
```

Similar or dependent issues:
* #4817 , #4814 
I'm assuming this is why the above issues have no "output".

## Additional Data
I've noticed in [PluginManager](https://github.com/serverless/serverless/blob/master/lib/classes/PluginManager.js#L64), the following snippet:
```
    // ignore plugins that specify a different provider than the current one
    if (pluginProvider
      && (pluginProvider !== this.serverless.service.provider.name)) {
      return;
    }
```
The issue is the provider variables have not been loaded by this step, so I'm assuming the `aws` plugins are then ignored?

* ***Serverless Framework Version you're using***:
1.26.0 (master branch at time of posting)
* ***Operating System***:
Ubuntu 16.04

**Comment by erikerikson** [comment]
Yes, the [filtering of plugins in PluginManager](https://github.com/serverless/serverless/blob/master/lib/classes/PluginManager.js#L64) combines with the fact that [plugins are loaded in init](https://github.com/serverless/serverless/blob/master/lib/Serverless.js#L67) whereas [the service is populated in run](https://github.com/serverless/serverless/blob/master/lib/Serverless.js#L89) and [init is called before run](https://github.com/serverless/serverless/blob/master/bin/serverless#L42).

This definitely represents a hidden dissonance with the mental model of engineers in using variables within `serverless.[json||yml||yaml]` and the returned content of `serverless.js`.  One challenge of not doing things in this order is that the proper population of a service can depend upon the provider that the service targets.  Not sure whether this one should be marked as a "bug" as "working as expected".

**Comment by MacMcIrish** [comment]
@erikerikson Thanks for replying. I think the misunderstanding comes from how the [second example](https://serverless.com/framework/docs/providers/aws/guide/variables#recursively-reference-properties) for serverless variables uses the provider as the target, it's not really clear that the provider name has any kind of restriction on it, or any of the provider variables really.

If it's hard baked into the design to not support variables for this `provider.name`, would it make sense to include a warning in the docs that the provider name cannot be a variable?

**Comment by erikerikson** [comment]
That at least seems a reasonable short term mitigation.

I think there's a long term argument for provider being legacy that disappears.  Particularly for the rare but important cross provider use cases.

Given the declaration of plugins in serverless.yml, it could seem that the resolution order is somewhat backwards.  That said, it's certainly legacy and the way of it for the time being.  Switching would also have the potential impact of disallowing plugin's involvement in variable resolution which is desirable for a few rather sticky scenarios involving conflicts.

**Comment by medikoo** [comment]
Fixed with #8987
